src/Asm89/Stack/CorsService.php

```diff
@@ -72,134 +67,159 @@ public function isPreflightRequest(Request $request)
-            : implode(', ', $this->options['allowedHeaders']);
-        $response->headers->set('Access-Control-Allow-Headers', $allowHeaders);
+        return false;
+    }
 
-        $response->setStatusCode(204);
+    public function addActualRequestHeaders(Response $response, Request $request)
+    {
+        $this->configureAllowedOrigin($response, $request);
+
+        $this->configureAllowCredentials($response, $request);
+
+        $this->configureExposedHeaders($response, $request);
 
         return $response;
     }
 
-    private function checkPreflightRequestConditions(Request $request)
+    private function configureAllowedOrigin(Response $response, Request $request)
     {
-        if (!$this->checkOrigin($request)) {
-            return $this->createBadRequestResponse(403, 'Origin not allowed');
+        if ($this->options['allowedOrigins'] === true && !$this->options['supportsCredentials']) {
+            // Safe+cacheable, allow everything
+            $response->headers->set('Access-Control-Allow-Origin', '*');
+        } elseif ($this->isSingleOriginAllowed()) {
+            // Single origins can be safely set
+            $response->headers->set('Access-Control-Allow-Origin', array_values($this->options['allowedOrigins'])[0]);
+        } else {
+            // For dynamic headers, check the origin first
+            if ($this->isOriginAllowed($request)) {
+                $response->headers->set('Access-Control-Allow-Origin', $request->headers->get('Origin'));
+            }
+
+            $this->varyHeader($response, 'Origin');
         }
+    }
 
-        if (!$this->checkMethod($request)) {
-            return $this->createBadRequestResponse(405, 'Method not allowed');
+    private function isSingleOriginAllowed()
+    {
+        if ($this->options['allowedOrigins'] === true || !empty($this->options['allowedOriginsPatterns'])) {
+            return false;
         }
 
-        $requestHeaders = array();
-        // if allowedHeaders has been set to true ('*' allow all flag) just skip this check
-        if ($this->options['allowedHeaders'] !== true && $request->headers->has('Access-Control-Request-Headers')) {
-            $headers        = strtolower($request->headers->get('Access-Control-Request-Headers'));
-            $requestHeaders = array_filter(explode(',', $headers));
+        return count($this->options['allowedOrigins']) === 1;
+    }
 
-            foreach ($requestHeaders as $header) {
-                if (!in_array(trim($header), $this->options['allowedHeaders'])) {
-                    return $this->createBadRequestResponse(403, 'Header not allowed');
-                }
+    private function configureAllowedMethods(Response $response, Request $request)
+    {
+        if ($this->options['allowedMethods'] === true) {
+            if ($this->options['supportsCredentials']) {
+                $allowMethods = 'GET, HEAD, PUT, PATCH, POST, DELETE';
```

Comment:
What about custom/non-standard methods?

Reply:
And this isn't even all of the standard ones:
https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods
but yeah I think it would be better to `Vary` this like `CorsService::configureAllowedHeaders()`

Reply:
Yeah perhaps so. But maybe the list of methods as an array would make more sense of a default config.

Reply:
I think it would be better to add some sort of configuration exception than do this, something like: https://github.com/davidbarratt/stack-cors/blob/00c8558ff726c5b369a6981a108313d1386db12b/src/Asm89/Stack/CorsService.php#L65-L69

or as an example:
```php
if ($options['supportsCredentials'] === true && $options['allowedMethods'] === true) {
  // This is not allowed.
  // @see https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Methods
  throw new \LogicException("Cannot set supportsCredentials to true and allowedMethods to ['*']");
}
```
That will force the user to specify a static whitelist of allowed headers, rather than assuming a specific list that may not include ones they are using.

But, I think it would be better to just `Vary` by the `Access-Control-Request-Method` since that's allows them to actually do what they are asking for (though, whith great cost).

Reply:
I've updated it to just use the Vary header.